### PR TITLE
tools/ci: Modify cibuild.sh to support new platforms

### PR DESCRIPTION
**Pull request by simbit18**
## Summary
This is an idea for tweaking how we build on GitHub and test locally. This will make it easier to maintain and make it easier to develop on new platforms. 

The proposed solution is based on the following changes:

- **cibuild.sh** -> Only detect platforms.
- **platforms** ->  Added a folder with scripts for each platform (installing toolchains, tools, etc.). 
                               darwin.sh, msys2.sh, linux.sh (Without the package management commands) and ubuntu.sh
 
Why each distribution has different packages (apt, apk, dnf, pacman, pkg) and each distribution has different versions of programs and  their dependencies.

## Impact
There should be no impact.

## Testing
We tested on the local platform MSYS2 and GitHub

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@simbit18 should we still keep the common code in cibuild.sh, instead duplicating in every shell script.

File: tools/ci/platforms/ubuntu.sh [hunk 1]
```diff
@@ -0,0 +1,543 @@
+#!/usr/bin/env bash
+############################################################################
+# tools/ci/platforms/ubuntu.sh
```
Comment (pkarashchenko) [inline tools/ci/platforms/ubuntu.sh:3]: Just for my understanding: what is the delta between `tools/ci/platforms/linux.sh` and `tools/ci/platforms/ubuntu.sh`?

File: tools/ci/platforms/linux.sh [hunk 2]
```diff
@@ -0,0 +1,541 @@
+#!/usr/bin/env bash
+
```
Comment (pkarashchenko) [inline tools/ci/platforms/linux.sh:2]:
```suggestion
############################################################################
 # tools/ci/platforms/linux.sh
```

File: tools/ci/platforms/darwin.sh
[the remark this answers is not in the document]
Comment (simbit18) [reply tools/ci/platforms/darwin.sh:1]: done

File: tools/ci/platforms/linux.sh (hunk 2 above)
Comment (simbit18) [reply tools/ci/platforms/linux.sh]: done

File: tools/ci/platforms/ubuntu.sh (hunk 1 above)
Comment (simbit18) [reply tools/ci/platforms/ubuntu.sh:3]: It is currently the same. This is a base for Linux distros that do not have the /etc/os-release file.

**Comment by simbit18** [comment]
@xiaoxiang781216 @pkarashchenko  Sorry for delayed reply. I returned from my vacation today.

File: tools/ci/platforms/ubuntu.sh (hunk 1 above)
Comment (pkarashchenko) [reply tools/ci/platforms/ubuntu.sh:3]: Maybe let's rename `linux.sh` to `debian.sh` as it early relies on `apt-get` and I think will not work for Fedora?

Comment (simbit18) [reply tools/ci/platforms/ubuntu.sh:3]: Recent Debian and Fedora distros have the /etc/os-release file and are already present in the find_platform function.

[function -> find_platform](https://github.com/apache/nuttx/blob/81d163710f2bcedc196d6890bf730fd25e92abe3/tools/ci/cibuild.sh#L46)

All that is missing is to create the script file (debian.sh and fedora.sh) with with the correct tool packages and toolchains.
If the /etc/os-release file is missing you search platform via "uname -s" (darwin, cygwin etc.).  So linux.sh script is for all those distros dated or without  /etc/os-release file that you don't know package manager. Maybe it is better to delete from linux.sh the apt-get packages and leave only the toolchains downloads.
What do you think ?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 @pkarashchenko Sorry for delayed reply. I returned from my vacation today.

there are some dup in each shell script, why not keep cibuild.sh for the common code and source the platform shell script by uname output?

**Comment by simbit18** [comment]
Because the separate script is easy to maintain and makes it easier to develop on new platforms. A script with different packages (apt, apk, dnf, pacman and pkg) creates complicated testing and generates easy errors. /etc/os-release file allows us to detect linux distro and other data.

**Comment by xiaoxiang781216** [comment]
but the follow code is same in all platform:
```
set -e
set -o xtrace

WD=$(cd "$(dirname "$0")" && pwd)
WORKSPACE=$(cd "${WD}"/../../../../ && pwd -P)
nuttx=${WORKSPACE}/nuttx
apps=${WORKSPACE}/apps
tools=${WORKSPACE}/tools
EXTRA_PATH=

function add_path {
  PATH=$1:${PATH}
  EXTRA_PATH=$1:${EXTRA_PATH}
}
...
function usage {
  echo ""
  echo "USAGE: $0 [-i] [-s] [-c] [-*] <testlist>"
  echo "       $0 -h"
  echo ""
  echo "Where:"
  echo "  -i install tools"
  echo "  -s setup repos"
  echo "  -c enable ccache"
  echo "  -* support all options in testbuild.sh"
  echo "  -h will show this help test and terminate"
  echo "  <testlist> select testlist file"
  echo ""
  exit 1
}

function enable_ccache {
  export CCACHE_DIR="${tools}"/ccache
}

function setup_repos {
  pushd .
  if [ -d "${nuttx}" ]; then
    cd "${nuttx}"; git pull
  else
    git clone https://github.com/apache/nuttx.git "${nuttx}"
    cd "${nuttx}"
  fi
  git log -1

  if [ -d "${apps}" ]; then
    cd "${apps}"; git pull
  else
    git clone https://github.com/apache/nuttx-apps.git "${apps}"
    cd "${apps}"
  fi
  git log -1
  popd
}

if [ -z "$1" ]; then
   usage
fi

while [ -n "$1" ]; do
  case "$1" in
  -h )
    usage
    ;;
  -i )
    install_tools
    ;;
  -c )
    enable_ccache
    ;;
  -s )
    setup_repos
    ;;
  -* )
    options+="$1 "
    ;;
  * )
    builds=( "$@" )
    break
    ;;
  esac
  shift
done

run_builds
```

**Comment by pkarashchenko** [comment]
I do not really understand the benefits of this change

File: tools/ci/platforms/darwin.sh [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+  echo ""
+  echo "USAGE: $0 [-i] [-s] [-c] [-*] <testlist>"
+  echo "       $0 -h"
+  echo ""
+  echo "Where:"
+  echo "  -i install tools"
+  echo "  -s setup repos"
+  echo "  -c enable ccache"
+  echo "  -* support all options in testbuild.sh"
+  echo "  -h will show this help test and terminate"
```
Comment (hartmannathan) [inline tools/ci/platforms/darwin.sh:322]:
```suggestion
  echo "  -h will show this help text and terminate"
```

**Review by hartmannathan — COMMENTED** [review]
Is it possible to use sh instead of bash and remove bashisms? On some systems bash is actually dash or some other implementation.

ShellCheck might be helpful. See: https://www.shellcheck.net/ and https://github.com/koalaman/shellcheck

File: tools/ci/platforms/ubuntu.sh [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+  echo ""
+  echo "USAGE: $0 [-i] [-s] [-c] [-*] <testlist>"
+  echo "       $0 -h"
+  echo ""
+  echo "Where:"
+  echo "  -i install tools"
+  echo "  -s setup repos"
+  echo "  -c enable ccache"
+  echo "  -* support all options in testbuild.sh"
+  echo "  -h will show this help test and terminate"
```
[the remark this answers is not in the document]
Comment (simbit18) [reply tools/ci/platforms/ubuntu.sh]: done

**Comment by simbit18** [comment]
@xiaoxiang781216  Kept the common code in cibuild.sh.

@hartmannathan  thanks for the link to the ShellCheck tool.

**Comment by simbit18** [comment]
@xiaoxiang781216 @hartmannathan  @pkarashchenko 

Fixed compilation problems and Kept the common code in cibuild.sh.

File: tools/ci/platforms/darwin.sh [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+# Prerequisites for macOS
+#  - Xcode (cc, etc)
+#  - homebrew
+#  - autoconf
+#  - wget
+
+set -e
+set -o xtrace
+
+WD=$(cd "$(dirname "$0")" && pwd)
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh:32]: could be shared too

File: tools/ci/platforms/darwin.sh [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+WORKSPACE=$(cd "${WD}"/../../../../ && pwd -P)
+nuttx=${WORKSPACE}/nuttx
+apps=${WORKSPACE}/apps
+tools=${WORKSPACE}/tools
+EXTRA_PATH=
+
+function add_path {
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh:37]: ditto

File: tools/ci/platforms/darwin.sh [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+    rm ${wasmbasefile}.tar.gz
+  fi
+
+  export WASI_SDK_PATH="${tools}/wasi-sdk"
+
+  command ${WASI_SDK_PATH}/bin/clang --version
+  command wamrc --version
+}
+
+function usage {
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh]: ditto

File: tools/ci/platforms/darwin.sh [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+  done
+  popd
+
+  if [ -d "${CCACHE_DIR}" ]; then
+    setup_links
+  fi
+  echo PATH="${EXTRA_PATH}"/"${PATH}" > "${tools}"/env.sh
+}
+
+function setup_repos {
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh]: shareable

File: tools/ci/platforms/darwin.sh [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+    cd "${apps}"; git pull
+  else
+    git clone https://github.com/apache/nuttx-apps.git "${apps}"
+    cd "${apps}"
+  fi
+  git log -1
+  popd
+}
+
+function run_builds {
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh]: ditto

File: tools/ci/platforms/darwin.sh [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+    # Print a summary of configuration and statistics counters
+    ccache -s
+  fi
+}
+
+if [ -z "$1" ]; then
+   usage
+fi
+
+while [ -n "$1" ]; do
```
Comment (xiaoxiang781216) [inline tools/ci/platforms/darwin.sh]: could be shared

**Comment by simbit18** [comment]
@xiaoxiang781216 done

@hartmannathan Removed bashisms from scripts darwin.sh, linux.sh, msys2.sh and ubuntu.sh.

**Review by hartmannathan — APPROVED** [review]
LGTM! Thanks for making it work without the bashisms!

**OPTIONAL** suggestion:

I noticed that sometimes we're using wget, sometimes curl.

Here's a trick (I got it from here: https://svn.apache.org/viewvc/subversion/trunk/get-deps.sh?revision=1905955&view=markup#l59 ) to detect which program is available, in order of preference wget, curl, fetch:

```
HTTP_FETCH=
[ -z "$HTTP_FETCH" ] && type wget  >/dev/null 2>&1 && HTTP_FETCH="wget -q -nc"
[ -z "$HTTP_FETCH" ] && type curl  >/dev/null 2>&1 && HTTP_FETCH="curl -sOL"
[ -z "$HTTP_FETCH" ] && type fetch >/dev/null 2>&1 && HTTP_FETCH="fetch -q"
```

And later in the script you can use `$HTTP_FETCH` instead of hard-coding the program name. For example:

```
$HTTP_FETCH https://archive.apache.org/dist/serf/$SERF.tar.bz2
```

Caveat: In a few places we're calling curl with `-L` and `-s` but not `-O` because we want to control the output filename. The trick above won't work because it calls curl with `-sOL`. A possible workaround is to get the file with `-sOL` and then rename it to the desired name.

Like I said above, this is an **OPTIONAL** suggestion!! We can leave the scripts as-is.

**Comment by simbit18** [comment]
@hartmannathan Perhaps it is preferable that we just use cURL tool. I have seen that it is required in both NuttX core and apps.

some examples:
NuttX

[Config.mk](https://github.com/apache/nuttx/blob/954bffcbd771516952990f0c6117bbb973564425/tools/Config.mk#L559)
all arch/xtensa for bootloader-esp32X.bin and partition-table-esp32X.bin  [Bootloader.mk](https://github.com/apache/nuttx/blob/954bffcbd771516952990f0c6117bbb973564425/arch/xtensa/src/esp32/Bootloader.mk#L197)
[newlib](https://github.com/apache/nuttx/blob/954bffcbd771516952990f0c6117bbb973564425/libs/libm/newlib/Make.defs#L33)
arch/nrf{52|53|91 [nrf91](https://github.com/apache/nuttx/blob/954bffcbd771516952990f0c6117bbb973564425/arch/arm/src/nrf91/Make.defs#L90)

Apps

[lua](https://github.com/apache/nuttx-apps/blob/7ad4a3dcb7b68981887e619d70d17f784283517f/interpreters/lua/Makefile#L70)
[wasm3](https://github.com/apache/nuttx-apps/blob/7ad4a3dcb7b68981887e619d70d17f784283517f/interpreters/wasm3/Makefile#L62)

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by hartmannathan** [comment]
> @hartmannathan Perhaps it is preferable that we just use cURL tool. I have seen that it is required in both NuttX core and apps.

In that case, yes, it is probably best to standardize on curl throughout the codebase.

I see that this PR is merged. Great job and a big THANK YOU!

**Comment by xiaoxiang781216** [comment]
> > @hartmannathan Perhaps it is preferable that we just use cURL tool. I have seen that it is required in both NuttX core and apps.
> 
> In that case, yes, it is probably best to standardize on curl throughout the codebase.
> 
> I see that this PR is merged. Great job and a big THANK YOU!

Yes, all download in apps folder converts to curl, it will be great that we continue to switch wget to curl in nuttx folder.

**Comment by masayuki2009** [comment]
@simbit18 

Please squash the commit messages when you send a PR next time.

```
commit fd86cf70cc9e4c7645686e66ce2a894642f3fcbb (HEAD -> master, upstream/master)
Author: simbit18 <101105604+simbit18@users.noreply.github.com>
Date:   Thu Feb 8 17:33:44 2024 +0100

    tools/ci: Modify cibuild.sh to adapt to new platforms.

    cibuild.sh -> modify scripts
    platforms -> new folder +scripts
    add header

    tools/ci/platforms: Set the execute permissions on the updated files.

    tools/ci/platforms/linux.sh: fix header

    Corrected Apache Foundation copyright header.

    tools/ci: Modify cibuild.sh to adapt to new platforms.

    cibuild.sh -> modify scripts
    platforms -> new folder +scripts
    add header

    tools/ci/platforms: Set the execute permissions on the updated files.

    tools/ci: Modify cibuild.sh to adapt to new platforms.

    cibuild.sh -> modify scripts
    platforms -> new folder +scripts
    add header

    tools/ci/platforms: Set the execute permissions on the updated files.

    tools/ci: Kept the common code in cibuild.sh.

    Kept the common code in cibuild.sh.

    tools/ci: Storing all Environment Variables in env.sh script.

    Added in cibuild.sh this command "source "${CIWORKSPACE}"/tools/env.sh" to activate installed tools.

    tools/ci: shared the tools path in cibuild.sh

    Added in cibuild.sh enviroment variable "NUTTXTOOLS" for installed tools.

    Removed bashisms from scripts darwin.sh, linux.sh, msys2.sh and ubuntu.sh.

    tools/ci: Fixed build problems with macOS

    Fix Error: /Users/runner/work/nuttx/nuttx/sources/tools/bloaty-src/build is not a directory
```

**Pull request merged**